Notebooks/06-End_to_End.ipynb

```diff
@@ -0,0 +1,699 @@
+    "    'trip_distance > 0 and trip_distance < 500',\n",
+    "    'not (trip_distance > 50 and fare_amount < 50)',\n",
+    "    'not (trip_distance < 10 and fare_amount > 300)',\n",
+    "    'not dropoff_datetime <= pickup_datetime'\n",
+    "]\n",
+    "taxi_df = taxi_df.query(' and '.join(query_frags))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# reset_index and drop index column\n",
+    "taxi_df = taxi_df.reset_index(drop=True)\n",
+    "taxi_df.head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Adding Interesting Features\n",
+    "\n",
+    "Dask & cuDF provide standard DataFrame operations, but also let you run \"user defined functions\" on the underlying data. Here we use [dask.dataframe's map_partitions](https://docs.dask.org/en/latest/dataframe-api.html#dask.dataframe.DataFrame.map_partitions) to apply user defined python function on each DataFrame partition.\n",
+    "\n",
+    "We'll use a Haversine Distance calculation to find total trip distance, and extract additional useful variables from the datetime fields."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "## add features\n",
+    "\n",
+    "taxi_df['hour'] = taxi_df['pickup_datetime'].dt.hour\n",
+    "taxi_df['year'] = taxi_df['pickup_datetime'].dt.year\n",
+    "taxi_df['month'] = taxi_df['pickup_datetime'].dt.month\n",
+    "taxi_df['day'] = taxi_df['pickup_datetime'].dt.day\n",
+    "taxi_df['day_of_week'] = taxi_df['pickup_datetime'].dt.weekday\n",
+    "taxi_df['is_weekend'] = (taxi_df['day_of_week']>=5).astype('int32')\n",
+    "\n",
+    "#calculate the time difference between dropoff and pickup.\n",
+    "taxi_df['diff'] = taxi_df['dropoff_datetime'].astype('int64') - taxi_df['pickup_datetime'].astype('int64')\n",
+    "taxi_df['diff']=(taxi_df['diff']/1000).astype('int64')\n",
+    "\n",
+    "taxi_df['pickup_latitude_r'] = taxi_df['pickup_latitude']//.01*.01\n",
+    "taxi_df['pickup_longitude_r'] = taxi_df['pickup_longitude']//.01*.01\n",
+    "taxi_df['dropoff_latitude_r'] = taxi_df['dropoff_latitude']//.01*.01\n",
+    "taxi_df['dropoff_longitude_r'] = taxi_df['dropoff_longitude']//.01*.01\n",
+    "\n",
+    "taxi_df = taxi_df.drop('pickup_datetime', axis=1)\n",
+    "taxi_df = taxi_df.drop('dropoff_datetime', axis=1)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def haversine_dist(df):\n",
+    "    df['h_distance']= cuspatial.haversine_distance(\n",
+    "        df['pickup_longitude'],\n",
+    "        df['pickup_latitude'],\n",
+    "        df['dropoff_longitude'],\n",
+    "        df['dropoff_latitude']\n",
+    "    )\n",
+    "    df['h_distance']= df['h_distance'].astype('float32')\n",
+    "    return df\n",
```

Comment:
Can we replace this distance to a native cupy kernel?   

The reason is that this makes a CPU to GPU comparison easier and removes `cuspatial` as a dependency.  

```python
def cudf_haversine_distance(lon1, lat1, lon2, lat2):

    lon1, lat1, lon2, lat2 = map(cp.radians, [lon1, lat1, lon2, lat2])

    newlon = lon2 - lon1
    newlat = lat2 - lat1

    haver_formula = cp.sin(newlat/2.0)**2 + cp.cos(lat1) * cp.cos(lat2) * cp.sin(newlon/2.0)**2

    dist = 2 * cp.arcsin(cp.sqrt(haver_formula ))
    km = 6367 * dist #6367 for distance in KM for miles use 3958
    return km


def haversine_dist(df):
    df['h_distance']= cudf_haversine_distance(
        df['pickup_longitude'],
        df['pickup_latitude'],
        df['dropoff_longitude'],
        df['dropoff_latitude']
    )
    df['h_distance']= df['h_distance'].astype('float32')
    return df
```


I can confirm that the run time and accuracy both remain the same with this function.

Reply:
We can, but I think there is value in highlighting that cuSpatial exists. Perhaps in your comparison notebook you could do this but in this one we keep it like this?

Reply:
That works too.

Reply:
Discussed offline. Keeping as is